File: 2_term/4_C#/4-1_C#/4-1_C#/ParserOfEquation.cs
using System;
using System.IO;

namespace _C
{
    public class ParserOfEquation
    {
        public static AriphTree ParseFile(string filename)
        {
            sw = new StreamReader(filename);
            string equation = sw.ReadToEnd();
            AriphTree tree = new AriphTree(equation[1]);
            for (int i = 2; i < equation.Length; i++)
            {
                if (equation[i] == '(')
                {
                    if (tree.IsLeftFree())
                    {
                        tree.СreateLeftElement(equation[i + 1]);
                        tree.DownLeft();
                    }
                    else
                    {
                        tree.СreateRightElement(equation[i + 1]);
                        tree.DownRight();
                    }
                }
                else if (equation[i] == ')')
                {
                    tree.Up();
                }
                else if (equation[i] >= '0' && equation[i] <= '9')
                {
                    if (tree.IsLeftFree())
                    {
                        tree.СreateLeftElement(equation[i]);
                    }
                    else
                    {
                        tree.СreateRightElement(equation[i]);
                    }
                }
            }
            tree.SetDefault();
            return tree;
        }

        private static StreamReader sw;
    }
}

File: 2_term/6/6-1/parserOfEquation.cpp
#include "parserOfEquation.h"
#include  <fstream>
#include <QFile>
#include <QTextStream>
using namespace std;

AriphTree *ParserOfEquation::ParseFile(QFile *file)
{
    file->open(QIODevice::ReadOnly);
    QTextStream in(file);
    QString equation = in.readLine();
    if (!file->exists())
    {
        return nullptr;
    }
    AriphTree *tree = new AriphTree(equation[1]);
    for (unsigned int i = 2; i < equation.length(); i++)
    {
        if (equation[i] == '(')
        {
            if (tree->IsLeftFree())
            {
                tree->CreateLeftElement(equation[i + 1]);
                tree->DownLeft();
            }
            else
            {
                tree->CreateRightElement(equation[i + 1]);
                tree->DownRight();
            }
        }
        else if (equation[i] == ')')
        {
            tree->Up();
        }
        else if (equation[i].isDigit())
        {
            if (tree->IsLeftFree())
            {
                tree->CreateLeftElement(equation[i]);
            }
            else
            {
                tree->CreateRightElement(equation[i]);
            }
        }
    }
    tree->SetDefault();
    return tree;
}
